File: core/src/nl/t64/game/rpg/screens/ScreenUI.java
package nl.t64.game.rpg.screens;

import com.badlogic.gdx.scenes.scene2d.Stage;
import com.badlogic.gdx.scenes.scene2d.ui.Window;
import lombok.AllArgsConstructor;
import nl.t64.game.rpg.screens.inventory.HeroesTable;
import nl.t64.game.rpg.screens.inventory.WindowSelector;
import nl.t64.game.rpg.screens.inventory.equipslot.EquipSlotsTables;
import nl.t64.game.rpg.screens.inventory.inventoryslot.InventorySlotsTable;
import nl.t64.game.rpg.screens.inventory.itemslot.ItemSlot;
import nl.t64.game.rpg.screens.inventory.tooltip.BaseTooltip;
import nl.t64.game.rpg.screens.shop.ShopSlotsTable;

import java.util.List;


@AllArgsConstructor
public abstract class ScreenUI {

    protected final Window equipWindow;

    protected final EquipSlotsTables equipSlotsTables;
    protected final InventorySlotsTable inventorySlotsTable;
    protected final HeroesTable heroesTable;

    protected final List<WindowSelector> tableList;
    protected int selectedTableIndex;

    public EquipSlotsTables getEquipSlotsTables() {
        return equipSlotsTables;
    }

    public InventorySlotsTable getInventorySlotsTable() {
        return inventorySlotsTable;
    }

    public ShopSlotsTable getShopSlotsTable() {
        throw new IllegalCallerException("ShopSlotsTable not implemented here.");
    }

    public void updateSelectedHero(Runnable updateHero) {
        getSelectedTable().deselectCurrentSlot();

        int oldCurrentIndex = equipSlotsTables.getIndexOfCurrentSlot();
        equipWindow.getChild(1).remove();
        updateHero.run();
        equipWindow.add(equipSlotsTables.getCurrentEquipTable());
        equipSlotsTables.setCurrentByIndex(oldCurrentIndex);

        setFocusOnSelectedTable();
        getSelectedTable().selectCurrentSlot();
    }

    public void selectPreviousTable() {
        getSelectedTable().hideTooltip();
        getSelectedTable().deselectCurrentSlot();
        selectedTableIndex--;
        if (selectedTableIndex < 0) {
            selectedTableIndex = tableList.size() - 1;
        }
        setFocusOnSelectedTable();
        getSelectedTable().selectCurrentSlot();
    }

    public void selectNextTable() {
        getSelectedTable().hideTooltip();
        getSelectedTable().deselectCurrentSlot();
        selectedTableIndex++;
        if (selectedTableIndex >= tableList.size()) {
            selectedTableIndex = 0;
        }
        setFocusOnSelectedTable();
        getSelectedTable().selectCurrentSlot();
    }

    public void toggleTooltip() {
        ItemSlot currentSlot = getSelectedTable().getCurrentSlot();
        BaseTooltip currentTooltip = getSelectedTable().getCurrentTooltip();
        currentTooltip.toggle(currentSlot);
    }

    public void toggleCompare() {
        ItemSlot currentSlot = getSelectedTable().getCurrentSlot();
        BaseTooltip currentTooltip = getSelectedTable().getCurrentTooltip();
        currentTooltip.toggleCompare(currentSlot);
    }

    public void unloadAssets() {
        heroesTable.disposePixmapTextures();
    }

    protected void setFocusOnSelectedTable() {
        getSelectedTable().setKeyboardFocus(getStage());
        getStage().draw();
    }

    protected WindowSelector getSelectedTable() {
        return tableList.get(selectedTableIndex);
    }

    private Stage getStage() {
        return equipWindow.getStage();
    }

}

File: core/src/nl/t64/game/rpg/screens/ScreenUI.kt
package nl.t64.game.rpg.screens

import com.badlogic.gdx.scenes.scene2d.Stage
import com.badlogic.gdx.scenes.scene2d.ui.Window
import nl.t64.game.rpg.screens.inventory.HeroesTable
import nl.t64.game.rpg.screens.inventory.WindowSelector
import nl.t64.game.rpg.screens.inventory.equipslot.EquipSlotsTables
import nl.t64.game.rpg.screens.inventory.inventoryslot.InventorySlotsTable
import nl.t64.game.rpg.screens.shop.ShopSlotsTable


abstract class ScreenUI(
    protected val equipWindow: Window,
    val equipSlotsTables: EquipSlotsTables,
    val inventorySlotsTable: InventorySlotsTable,
    protected val heroesTable: HeroesTable,
    private val tableList: List<WindowSelector>,
    private var selectedTableIndex: Int
) {
    private val stage: Stage get() = equipWindow.stage

    open fun getShopSlotsTable(): ShopSlotsTable {
        throw IllegalStateException("ShopSlotsTable not implemented here.")
    }

    fun updateSelectedHero(updateHero: Runnable) {
        getSelectedTable().deselectCurrentSlot()

        val oldCurrentIndex = equipSlotsTables.getIndexOfCurrentSlot()
        equipWindow.getChild(1).remove()
        updateHero.run()
        equipWindow.add(equipSlotsTables.getCurrentEquipTable())
        equipSlotsTables.setCurrentByIndex(oldCurrentIndex)

        setFocusOnSelectedTable()
        getSelectedTable().selectCurrentSlot()
    }

    fun selectPreviousTable() {
        getSelectedTable().hideTooltip()
        getSelectedTable().deselectCurrentSlot()
        selectedTableIndex--
        if (selectedTableIndex < 0) {
            selectedTableIndex = tableList.size - 1
        }
        setFocusOnSelectedTable()
        getSelectedTable().selectCurrentSlot()
    }

    fun selectNextTable() {
        getSelectedTable().hideTooltip()
        getSelectedTable().deselectCurrentSlot()
        selectedTableIndex++
        if (selectedTableIndex >= tableList.size) {
            selectedTableIndex = 0
        }
        setFocusOnSelectedTable()
        getSelectedTable().selectCurrentSlot()
    }

    fun toggleTooltip() {
        val currentSlot = getSelectedTable().getCurrentSlot()
        val currentTooltip = getSelectedTable().getCurrentTooltip()
        currentTooltip.toggle(currentSlot)
    }

    fun toggleCompare() {
        val currentSlot = getSelectedTable().getCurrentSlot()
        val currentTooltip = getSelectedTable().getCurrentTooltip()
        currentTooltip.toggleCompare(currentSlot)
    }

    fun unloadAssets() {
        heroesTable.disposePixmapTextures()
    }

    protected fun setFocusOnSelectedTable() {
        getSelectedTable().setKeyboardFocus(stage)
        stage.draw()
    }

    protected fun getSelectedTable(): WindowSelector {
        return tableList[selectedTableIndex]
    }

}
